apps/workflows/jobs/spend-management.ts

```diff
@@ -0,0 +1,269 @@
+        where: (table, { isNotNull, isNull, not, eq, and }) =>
+          and(
+            isNotNull(table.stripeCustomerId),
+            isNotNull(table.subscriptions),
+            not(eq(table.plan, "free")),
+            isNull(table.deletedAt),
+          ),
+        with: {
+          budgets: true,
+        },
+      }),
+    );
+
+    workspaces = workspaces.filter(
+      (ws) =>
+        ws.subscriptions &&
+        Object.keys(ws.subscriptions).length > 0 &&
+        ws.budgets.filter(isActiveBudget).length > 0,
+    );
+
+    io.logger.info(`found ${workspaces.length} workspaces`, workspaces);
+
+    const t = new Date();
+    t.setUTCMonth(t.getUTCMonth() - 1);
+    const year = t.getUTCFullYear();
+    const month = t.getUTCMonth() + 1; // months are 0 indexed
+
+    // TODO: Find a way to pipe Tinybird for a given set of workspaceIds.
+    const activeKeys = await io.runTask("list involved workspaces active keys", async () =>
+      Promise.all(
+        workspaces.map(async (workspace) =>
+          tinybird
+            .activeKeys({
+              workspaceId: workspace.id,
+              year,
+              month,
+            })
+            .then((res) => ({
+              activeKeys: res.data.at(0)?.keys ?? 0,
+              workspaceId: workspace.id,
+            })),
+        ),
+      ),
+    );
+
+    // TODO: Find a way to pipe Tinybird for a given set of workspaceIds.
+    const verifications = await io.runTask("list involved workspaces key verifications", async () =>
+      Promise.all(
+        workspaces.map(async (workspace) =>
+          tinybird
+            .verifications({
+              workspaceId: workspace.id,
+              year,
+              month,
+            })
+            .then((res) => ({
+              verifications: res.data.at(0)?.success ?? 0,
+              workspaceId: workspace.id,
+            })),
+        ),
+      ),
+    );
+
+    io.logger.info("estimate workspaces current billing");
+    const currentUsages = workspaces.map((workspace) => {
+      const usedActiveKeys =
+        activeKeys.find((usage) => usage.workspaceId === workspace.id)?.activeKeys || 0;
+      const usedVerifications =
+        verifications.find((usage) => usage.workspaceId === workspace.id)?.verifications || 0;
+
+      let currentPrice = 0;
+      if (workspace.subscriptions?.plan) {
+        const cost = parseFloat(workspace.subscriptions.plan.cents);
+        currentPrice += cost;
+      }
+      if (workspace.subscriptions?.support) {
+        const cost = parseFloat(workspace.subscriptions.support.cents);
+        currentPrice += cost;
+      }
+      if (workspace.subscriptions?.activeKeys) {
+        const cost = calculateTieredPrices(
+          workspace.subscriptions.activeKeys.tiers,
+          usedActiveKeys,
+        );
+        // TODO: What to do on error? Skip billing? Assume zero?
+        if (!cost.err) {
+          currentPrice += cost.val.totalCentsEstimate;
+        }
```

Comment:
In the outlier case that the calculation of current usage has an error, should it skip the billing notification or assume it as zero?

Reply:
it should probably just crash, then we can debug it in trigger's dashboard